File: services/ml-service/app/instrumentation.py
"""
instrumentation.py — OpenTelemetry + Sentry initialisation for ml-service.

Call configure() once at application startup (before FastAPI is created).

Environment variables:
  OTEL_SERVICE_NAME              ml-service (default)
  OTEL_EXPORTER_OTLP_ENDPOINT    http://otel-collector:4318 (default)
  OTEL_TRACES_SAMPLER            parentbased_traceidratio (default)
  OTEL_TRACES_SAMPLER_ARG        1.0 (100% in dev; reduce to 0.05 in prod)
  SENTRY_DSN                     (optional)
  DEPLOY_ENV                     production | staging | development (default)
"""

from __future__ import annotations

import logging
import os

logger = logging.getLogger("instrumentation")

_SERVICE_NAME  = os.getenv("OTEL_SERVICE_NAME", "ml-service")
_OTEL_ENDPOINT = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://otel-collector:4318")
_DEPLOY_ENV    = os.getenv("DEPLOY_ENV", "development")
_SAMPLE_RATE   = float(os.getenv("OTEL_TRACES_SAMPLER_ARG", "1.0"))

_configured = False


def configure() -> None:
    """
    Configure OpenTelemetry SDK and Sentry.
    Safe to call multiple times — only configures once.
    """
    global _configured
    if _configured:
        return
    _configured = True

    _init_otel()
    _init_sentry()


def _init_otel() -> None:
    try:
        from opentelemetry import trace
        from opentelemetry.sdk.resources import Resource, SERVICE_NAME, SERVICE_VERSION
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.sampling import ParentBasedTraceIdRatio
        from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
        from opentelemetry.sdk.trace.export import BatchSpanProcessor
        from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
        from opentelemetry.instrumentation.requests import RequestsInstrumentor
        from opentelemetry.instrumentation.sqlalchemy import SQLAlchemyInstrumentor
        from opentelemetry.instrumentation.psycopg2 import Psycopg2Instrumentor

        resource = Resource.create({
            SERVICE_NAME:    _SERVICE_NAME,
            SERVICE_VERSION: os.getenv("APP_VERSION", "0.0.1"),
            "deployment.environment": _DEPLOY_ENV,
        })

        provider = TracerProvider(
            resource=resource,
            sampler=ParentBasedTraceIdRatio(_SAMPLE_RATE),
        )

        exporter = OTLPSpanExporter(
            endpoint=f"{_OTEL_ENDPOINT}/v1/traces",
        )
        provider.add_span_processor(BatchSpanProcessor(exporter))
        trace.set_tracer_provider(provider)

        # Auto-instrument HTTP clients and DB drivers
        RequestsInstrumentor().instrument()
        Psycopg2Instrumentor().instrument()

        # FastAPI instrumentor is applied in main.py via FastAPIInstrumentor().instrument_app(app)
        # (must happen after app is created)

        logger.info(
            f"[instrumentation] OTel SDK configured — "
            f"service={_SERVICE_NAME} endpoint={_OTEL_ENDPOINT} sample_rate={_SAMPLE_RATE}"
        )

    except ImportError as exc:
        logger.warning(
            f"[instrumentation] opentelemetry packages not fully installed: {exc}. "
            "Install with: pip install opentelemetry-sdk opentelemetry-exporter-otlp-proto-http "
            "opentelemetry-instrumentation-fastapi opentelemetry-instrumentation-psycopg2"
        )


def _init_sentry() -> None:
    dsn = os.getenv("SENTRY_DSN")
    if not dsn:
        return

    try:
        import sentry_sdk
        from sentry_sdk.integrations.fastapi import FastApiIntegration
        from sentry_sdk.integrations.logging import LoggingIntegration
        import logging as _logging

        sentry_sdk.init(
            dsn=dsn,
            environment=_DEPLOY_ENV,
            traces_sample_rate=0.05 if _DEPLOY_ENV == "production" else 1.0,
            integrations=[
                FastApiIntegration(),
                LoggingIntegration(
                    level=_logging.WARNING,
                    event_level=_logging.ERROR,
                ),
            ],
        )
        logger.info("[instrumentation] Sentry SDK initialised")
    except ImportError:
        logger.warning(
            "[instrumentation] sentry-sdk not installed — Sentry disabled. "
            "Install with: pip install sentry-sdk[fastapi]"
        )


def instrument_fastapi(app) -> None:
    """
    Apply FastAPI OTel instrumentation to an existing FastAPI app instance.
    Call this after configure() and after the FastAPI app is constructed.
    """
    try:
        from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
        FastAPIInstrumentor.instrument_app(
            app,
            excluded_urls="/health,/metrics",
        )
        logger.info("[instrumentation] FastAPI instrumentor applied")
    except ImportError:
        pass

File: services/game-service/src/instrumentation.ts
/**
 * instrumentation.ts — OpenTelemetry SDK initialisation for game-service.
 *
 * Must be the FIRST import in src/index.ts (before Express, Prisma, etc.)
 * so that auto-instrumentation patches are applied before any module is loaded.
 *
 * Configured via environment variables:
 *   OTEL_SERVICE_NAME          game-service (default)
 *   OTEL_EXPORTER_OTLP_ENDPOINT http://otel-collector:4318 (default)
 *   OTEL_TRACES_SAMPLER        parentbased_traceidratio (default)
 *   OTEL_TRACES_SAMPLER_ARG    1.0 (default — 100% sampling in dev; lower in prod)
 *   SENTRY_DSN                 (optional) Sentry error tracking DSN
 *   NODE_ENV                   production | staging | development
 */

import { NodeSDK } from '@opentelemetry/sdk-node';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-http';
import { PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
import { Resource } from '@opentelemetry/resources';
import { SEMRESATTRS_SERVICE_NAME, SEMRESATTRS_SERVICE_VERSION, SEMRESATTRS_DEPLOYMENT_ENVIRONMENT } from '@opentelemetry/semantic-conventions';

const SERVICE_NAME    = process.env.OTEL_SERVICE_NAME    ?? 'game-service';
const OTEL_ENDPOINT   = process.env.OTEL_EXPORTER_OTLP_ENDPOINT ?? 'http://otel-collector:4318';
const DEPLOY_ENV      = process.env.NODE_ENV ?? 'development';
const SERVICE_VERSION = process.env.npm_package_version  ?? '0.0.1';

// ── Sentry (optional) ─────────────────────────────────────────────────────────

function initSentry(): void {
    const dsn = process.env.SENTRY_DSN;
    if (!dsn) return;

    try {
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const Sentry = require('@sentry/node') as typeof import('@sentry/node');
        Sentry.init({
            dsn,
            environment:  DEPLOY_ENV,
            release:      `${SERVICE_NAME}@${SERVICE_VERSION}`,
            tracesSampleRate: DEPLOY_ENV === 'production' ? 0.1 : 1.0,
            // Capture unhandled promise rejections and uncaught exceptions
            integrations: [
                Sentry.captureConsoleIntegration({ levels: ['error', 'warn'] }),
            ],
        });
        console.info('[instrumentation] Sentry initialised');
    } catch {
        console.warn('[instrumentation] @sentry/node not installed — Sentry disabled');
    }
}

// ── OpenTelemetry SDK ─────────────────────────────────────────────────────────

const traceExporter = new OTLPTraceExporter({
    url: `${OTEL_ENDPOINT}/v1/traces`,
});

const metricExporter = new OTLPMetricExporter({
    url: `${OTEL_ENDPOINT}/v1/metrics`,
});

const sdk = new NodeSDK({
    resource: new Resource({
        [SEMRESATTRS_SERVICE_NAME]:           SERVICE_NAME,
        [SEMRESATTRS_SERVICE_VERSION]:        SERVICE_VERSION,
        [SEMRESATTRS_DEPLOYMENT_ENVIRONMENT]: DEPLOY_ENV,
    }),
    traceExporter,
    metricReader: new PeriodicExportingMetricReader({
        exporter: metricExporter,
        exportIntervalMillis: 15_000,
    }),
    instrumentations: [
        getNodeAutoInstrumentations({
            // HTTP: auto-instrument Express routes
            '@opentelemetry/instrumentation-http': {
                ignoreIncomingRequestHook: (req) =>
                    req.url === '/health' || req.url === '/metrics',
            },
            // Prisma ORM — trace DB queries
            '@opentelemetry/instrumentation-prisma': { enabled: true },
            // Disable noisy fs instrumentation
            '@opentelemetry/instrumentation-fs': { enabled: false },
        }),
    ],
});

// ── Startup ───────────────────────────────────────────────────────────────────

try {
    sdk.start();
    initSentry();
    console.info(
        `[instrumentation] OTel SDK started — service=${SERVICE_NAME} env=${DEPLOY_ENV} endpoint=${OTEL_ENDPOINT}`
    );
} catch (err) {
    console.error('[instrumentation] OTel SDK failed to start:', err);
}

// Graceful shutdown: flush spans before process exits
process.on('SIGTERM', () => {
    sdk.shutdown()
        .then(() => console.info('[instrumentation] OTel SDK shut down'))
        .catch((err) => console.error('[instrumentation] OTel shutdown error:', err));
});
